### ISSUE TITLE
[BUG] changes to 'restart' in docker-compose.yml not taken into account?

### ISSUE THREAD

**Issue by gggeek** [docker/compose#11679]
### Description

I have set `restart: "on-failure:3"` for multiple services within a docker compose yml file, and tested that, after doing that, the containers were correctly started upon restarting the host vm, as well as when manually stopping then restarting the docker service.

I have then changed the compose yml file to `restart: "${COMPOSE_CONTAINERS_RESTART:-no}"`, and added `COMPOSE_CONTAINERS_RESTART=no` to the .env file.

When running `docker compose config`, I now see `restart: "no"` within the dumped information.

But the containers keep being restarted whenever the host is rebooted or the docker service is restarted, which is unexpected.

I wonder if the docker daemon actually stores somewhere the original value for the 'restart' element, and uses it later, regardless of the current value. In fact, I am not even sure if it is sensible from a logical pov to have that value come from an env var... Do I have to rebuild the containers from scratch?

Note: my usecase is to have multiple copies of the same docker-based stack running in parallel on the same host computer, of which only one has to be automatically restarted. The are multiple docker compose yml files in use, and they make heavy usage of env vars to allow the different copies to behave differently.

### Compose Version

```Text
Docker Compose version v2.25.0
```

### Docker Environment

```Text
Client: Docker Engine - Community
 Version:    26.0.0
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.13.1
    Path:     /usr/libexec/docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.25.0
    Path:     /usr/libexec/docker/cli-plugins/docker-compose

Server:
 Containers: 34
  Running: 0
  Paused: 0
  Stopped: 34
 Images: 127
 Server Version: 26.0.0
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: ae07eda36dd25f8a1b98dfbf587313b99c0190bb
 runc version: v1.1.12-0-g51d5e94
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: builtin
  cgroupns
 Kernel Version: 5.15.0-101-generic
 Operating System: Ubuntu 22.04.4 LTS
 OSType: linux
 Architecture: x86_64
 CPUs: 2
 Total Memory: 15.6GiB
 Name: ubuntu22
 ID: 6OZ2:4AQF:L7H3:IOX5:R5TP:OPSN:P5Q5:7VJ7:XCOH:YDU3:N5O4:4PM2
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Username: gggeek
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

**Comment by gggeek** [comment]
PS: further experiments seem to show that using the `--force-recreate` flag when starting the containers (via `docker compose up`) does actually make the docker daemon take into account the new value...

**Comment by ndeloof** [comment]
> I wonder if the docker daemon actually stores somewhere the original value for the 'restart' element

indeed. Compose.yaml is only used to create container, then the running resource config is all stored in docker engine. Changes you make to compose.yaml will have no effect until you run `compose up` again.

**Comment by gggeek** [comment]
If I get what you are saying, the changes in compose.yaml are not being picked up by restarting the host or the docker service, and that is expected behaviour, but they would be picked up as soon as `compose up` is run.

My testing seems to indicate instead that `compose up` is not enough, and `compose up --force-recreate` is needed.
If that is the case, I can probably live with the behaviour not being considered a bug, but it would be nice to have it documented.

I will do another round of tests...

**Comment by ndeloof** [comment]
Ran a quick test to confirm:

```console
$ cat compose.yaml 
services:
    one:
        image: alpine
        command: ping localhost
        restart: always
$ docker compose up -d
[+] Running 1/1
 ✔ Container machin-one-1  Started                                         0.1s 
 $ vi compose.yaml  
# edit  restart: no
 $ docker compose up -d
[+] Running 0/1
 ⠦ Container machin-one-1  Recreate                                        6.7s
```
container is recreated as compose detect config for service was modified
